rfcs/0050-merge-bot-for-maintainers.md

```diff
@@ -0,0 +1,116 @@
+---
+feature: merge-bot-for-maintainers
+start-date: 2019-08-17
+author: Frederik Rietdijk
+co-authors: (find a buddy later to help our with the RFC)
+shepherd-team: @aanderse, @globin, @grahamc, @worldofpeace
+shepherd-leader: @globin
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Allow maintainers of packages in Nixpkgs to submit their own changes
+when a trusted reviewer reviewed their change and other criteria are fullfilled.
+When the criteria are fullfilled a change can be merged using
+
+    @grahamcofborg merge
+
+# Motivation
+[motivation]: #motivation
+
+Nixpkgs is growing. Community members want to be able and should be able to help out
+with merging changes. Merging changes typically requires write permission, however,
+that gives the possibility for Nixpkgs-wide changes and decreases the security.
+
+By allowing maintainers of packages and modules to submit their own changes the
+load on members of the @nixpkgs-committers that have write permission group is
+reduced.
+
+A new group, trusted reviewers, is introduced to help out with the reviewing
+process to improve the reviewing culture by introducing peer-review and making
+the whole process more inclusive.
+
+# Detailed design
+[design]: #detailed-design
+
+Community members will be able to submit their changes using a bot after a
+positive review of both the maintainers of the expressions they modify
+("relevant maintainers") as well that of a trusted reviewer.
+
+## Scope
+
+- Only packages are considered. Modules are excluded because of how intertwined modules are.
+- Only `master` and `staging` branches are considered. In the future release branches may be included as well.
+
+## Trusted reviewers
+
+A new trusted reviewers group is introduced, called @nixpkgs-reviewers. This
+group is permission-wise positioned between @nixpkgs-maintainers and
+@nixpkgs-committers and can be considered a stepping stone towards becoming a
+member of @nixpkgs-committers and obtaining push permission.
+
+This new group is initially implemented with GitHub Teams. In the future this
+may be done differently, e.g. when narrower control lists are pursued.
+
+## Merge bot behavior
+
+The bot @grahamcofborg will support two new commands, `@grahamcofborg merge` and
+`@grahamcofborg stop` for respectively merging a PR and aborting a PR merge.
+
+### Merging a change
+
+The following requirements need to be fullfilled for the bot to be able to merge:
+1. Target branch of PR is `master` or `staging`.
+2. Review is green. A review is green when the relevant maintainers gave a positive review as well as a member of the trusted reviewers. Alternatively, a positive review of a committer is needed.
```

Comment:
I think that we don't need to wait for review for all changes. If there's a situation when package maintainer (or someone else, but package maintainer has approved change) just increase a version, then there's actually nothing to review, we can merge it automatically.

Technically algorithm will look like a:
1. Check that the only difference is attributes _**version**_ and _**sha256**_ sum.
2. Check that the attribute _**version**_ is increased.
3. Make sure that commit is signed (and have a valid signature) with the key identifier from maintainer-list.nix.